src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -181,45 +181,67 @@ public static class FrozenDictionary
                         }
                         else
                         {
-                            frozenDictionary = analysis.HashCount == 1
-                                ? new OrdinalStringFrozenDictionary_LeftJustifiedSingleChar<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex)
-                                : new OrdinalStringFrozenDictionary_LeftJustifiedSubstring<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount);
+                            if (analysis.HashCount == 1)
+                            {
+                                frozenDictionary = analysis.IgnoreCase
+                                    ? new OrdinalStringFrozenDictionary_LeftJustifiedSingleCharCaseInsensitive<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex)
+                                    : new OrdinalStringFrozenDictionary_LeftJustifiedSingleChar<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex);
+                            }
+                            else
+                            {
+                                frozenDictionary = analysis.IgnoreCase
+                                    ? new OrdinalStringFrozenDictionary_LeftJustifiedSubstringCaseInsensitive<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount)
+                                    : new OrdinalStringFrozenDictionary_LeftJustifiedSubstring<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount);
+                            }
                         }
                     }
                 }
                 else
                 {
-                    if (analysis.IgnoreCase)
+                    if (analysis.IgnoreCaseForHash)
                     {
-                        frozenDictionary = analysis.AllAsciiIfIgnoreCase
+                        Debug.Assert(analysis.IgnoreCase);
+                        frozenDictionary = analysis.AllAsciiIfIgnoreCaseForHash
                             ? new OrdinalStringFrozenDictionary_FullCaseInsensitiveAscii<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff)
                             : new OrdinalStringFrozenDictionary_FullCaseInsensitive<TValue>(keys, values, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff);
                     }
                     else
                     {
+                        // if (IgnoreCase) => Can only be true if there are no letters, thus case sensitive comparison still works here.
```

Comment:
Is there an assertion we could replace with this comment?

Reply:
I don't think so because the assertion would just be re-calculating what the `KeyAnalyzer` does.